pkg/tanka/find.go

```diff
@@ -2,123 +2,165 @@ package tanka
+	Selector    labels.Selector
+	Parallelism int
 }
 
 // FindEnvs returns metadata of all environments recursively found in 'path'.
 // Each directory is tested and included if it is a valid environment, either
 // static or inline. If a directory is a valid environment, its subdirectories
 // are not checked.
 func FindEnvs(path string, opts FindOpts) ([]*v1alpha1.Environment, error) {
-	// find all environments at dir
-	envs, errs := find(path, Opts{JsonnetOpts: opts.JsonnetOpts})
-	if errs != nil {
-		return envs, ErrParallel{errors: errs}
-	}
-
-	// optionally filter
-	if opts.Selector == nil || opts.Selector.Empty() {
-		return envs, nil
-	}
+	return findEnvsFromPaths([]string{path}, opts)
+}
 
-	filtered := make([]*v1alpha1.Environment, 0, len(envs))
-	for _, e := range envs {
-		if !opts.Selector.Matches(e.Metadata) {
-			continue
-		}
-		filtered = append(filtered, e)
-	}
+// FindEnvsFromPaths does the same as FindEnvs but takes a list of paths instead
+func FindEnvsFromPaths(paths []string, opts FindOpts) ([]*v1alpha1.Environment, error) {
+	return findEnvsFromPaths(paths, opts)
+}
 
-	return filtered, nil
+type findJsonnetFilesOut struct {
+	jsonnetFiles []string
+	err          error
 }
 
-func findErr(path string, err error) []error {
-	return []error{fmt.Errorf("%s:\n %w", path, err)}
+type findEnvsOut struct {
+	envs []*v1alpha1.Environment
+	err  error
 }
 
-// find implements the actual functionality described at 'FindEnvs'
-func find(path string, opts Opts) ([]*v1alpha1.Environment, []error) {
-	// try if this has envs
-	list, err := List(path, opts)
-	if err != nil &&
-		// expected when looking for environments
-		!errors.As(err, &jpath.ErrorNoBase{}) &&
-		!errors.As(err, &jpath.ErrorFileNotFound{}) {
-		return nil, findErr(path, err)
-	}
-	if len(list) != 0 {
-		// it has. don't search deeper
-		return list, nil
+func findEnvsFromPaths(paths []string, opts FindOpts) ([]*v1alpha1.Environment, error) {
+	if opts.Parallelism >= 0 {
+		opts.Parallelism = runtime.NumCPU()
 	}
 
-	stat, err := os.Stat(path)
-	if err != nil {
-		return nil, findErr(path, err)
+	log.Debug().Int("parallelism", opts.Parallelism).Int("paths", len(paths)).Msg("Finding Tanka environments")
+	startTime := time.Now()
+
+	// find all jsonnet files within given paths
+	pathChan := make(chan string, len(paths))
+	findJsonnetFilesChan := make(chan findJsonnetFilesOut)
+	for i := 0; i < opts.Parallelism; i++ {
+		go func() {
+			for path := range pathChan {
+				jsonnetFiles, err := jsonnet.FindFiles(path, nil)
+				var mainFiles []string
+				for _, file := range jsonnetFiles {
+					if filepath.Base(file) == jpath.DefaultEntrypoint {
+						mainFiles = append(mainFiles, file)
+					}
+				}
+				findJsonnetFilesChan <- findJsonnetFilesOut{jsonnetFiles: mainFiles, err: err}
+			}
+		}()
 	}
 
-	// if path is a file, don't search deeper
-	if !stat.IsDir() {
-		return nil, nil
+	// push paths to channel
+	var pathMap = map[string]bool{} // prevent duplicates
```

Comment:
Use a `struct{}` to not take any space, also suggest using `make` just for style.

This pattern of deduplicating and pushing to a channel is repeated further down too, so suggest refactoring to a helper function.

I don't see it in the previous code though - why is it needed now?

Reply:
Yeah, I think I'll just remove it. It's true that there was any deduplication before. That seems like a bug, but not one I'll correct within this refactoring